### PR TITLE
feat(fee_ticker): coingecko liquidity watcher

### PR DESCRIPTION
**Pull request by jurajpiar**
- changes uniswap to coingecko
- changes get_real_token_amount test in validator mod
- moves coin list requst to another function ✅
- calls that function in token watcher ✅
- passes result to this function ✅
- liquidity ticker changes from uniswap to coingecko

Solves:
[ROLLUP-88](https://rsklabs.atlassian.net/browse/ROLLUP-88)

### CODE REVIEW

**Review by antomor — CHANGES_REQUESTED** [review]

File: core/bin/zksync_api/src/fee_ticker/mod.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/mod.rs:214]: This is explained on the next line:
> // !! TODO: use config.ticker.coingecko_base_url.clone() once fee ticker and liquidity ticker are ran on the same mock server !!

Which points out (and reminds me at the same time) the reason why I was arguing for combining the dev-tickers. Otherwise, we could add a new config that would be something like `coinbase_liquidity_base_url`, or use the original `uniswap_base_url` at a cost of clarity. It's your call I guess @antomor.

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (jurajpiar) [reply docker/dev-liquidity-token-watcher/Dockerfile:2]: @franciscotobar, could you clarify:

- did you run it with version `rust:1.68`, or the previous value(`rust:latest`), or some other value?
- is the error copy-pasted? I'd expect it to fail for `sccache`, not `ccache` (the first is an extension of the latter, so this would suggest a new (to me) problem).
- did you clear the project and deleted all images before running it?

File: .gitignore
Comment (jurajpiar) [reply .gitignore]: removed

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (franciscotobar) [reply docker/dev-liquidity-token-watcher/Dockerfile:2]: I run it with the actual version, `FROM rust:1.59 as builder` that is on the file, and Im pretty sure that is `ccache`.

File: core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs:128]: The `plaform_id` refers to the chain_id, right?

File: core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs:114]: this function basically returns the `default_volume` from the config?

File: core/bin/zksync_api/src/fee_ticker/validator/mod.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/fee_ticker/validator/mod.rs:195]: what do you mean with online?

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/fee_ticker/validator/watcher.rs:20]: is this still valid?

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/fee_ticker/validator/watcher.rs:65]: what is the difference between platform and network?

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (franciscotobar) [reply docker/dev-liquidity-token-watcher/Dockerfile:2]: sorry I was using the wrong branch. I just tried and it works fine.

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (franciscotobar) [inline docker/dev-liquidity-token-watcher/Dockerfile:5]: why did we change on this PR, dont we have another PR for that?

**Review by franciscotobar — DISMISSED** [review]
LGTM

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (jurajpiar) [reply docker/dev-liquidity-token-watcher/Dockerfile:2]: thank you @franciscotobar, could you also try with `rust:latest` to verify/refute that it fails on M1, please?

File: core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/bin/dev-liquidity-token-watcher.rs:114]: correct

File: core/bin/zksync_api/src/fee_ticker/validator/mod.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/mod.rs:195]: 1. it is the original comment (and test), where I only changed uniswap to coingecko
2. to me this means that it shouldn't be run in the test suite, instead it should only be used to test the API connection, manually

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/watcher.rs:20]: it is not. thanks

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/watcher.rs:65]: hm... I guess if you don't know, then I failed to name it well. It refers to the coingecko platform for the current network (well, chain id).
would `find_coingecko_platform_for_current_chain_id` be clearer perhaps?

File: core/bin/zksync_api/src/fee_ticker/mod.rs
Comment (antomor) [reply core/bin/zksync_api/src/fee_ticker/mod.rs:214]: WDYT about adding a config param `coingecko_url`?

File: docker/dev-liquidity-token-watcher/Dockerfile
Comment (franciscotobar) [reply docker/dev-liquidity-token-watcher/Dockerfile:2]: it works with the `FROM rust:latest as builder`

File: core/bin/zksync_api/src/fee_ticker/mod.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/mod.rs:214]: this will be resolved in https://rsklabs.atlassian.net/browse/ROLLUP-182

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/watcher.rs]: done btw

File: core/bin/zksync_api/src/fee_ticker/validator/watcher.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/watcher.rs]: done

File: docker-compose.yml
Comment (antomor) [inline docker-compose.yml:29]:
```suggestion
      context: ./docker/dev-ticker/Dockerfile
```

**Review by antomor — CHANGES_REQUESTED** [review]

**Review by antomor — COMMENTED** [review]
@jurajpiar it took me a while reviewing it because I encountered some issues:
1. `zk init` failed because the `dev-ticker` image cannot be built directly (unless I run `zk docker build dev-ticker`, so I had to build it before and then use the previously built image. (I guess this isn't a big issue because we plan to publish it and use the previously published image instead of building it each time)
2. I've realised that the balance of the tokens we're using (wRBTC, RDOC, RIF) cannot be retrieved (either by the dapp or with metamask, adding the custom token); btw this issue isn't related to these changes.

File: docker-compose.yml
Comment (jurajpiar) [reply docker-compose.yml:29]: why?

Comment (antomor) [reply docker-compose.yml:30]: It looks the context is required, but we cannot get rid of `dockerfile` because we want to use `.` as context and `./docker/dev-ticker/Dockerfile` as docker file.

**Review by antomor — APPROVED** [review]
LGTM, since the issue that I noticed looks like unrelated to the PR. Thanks @jurajpiar

**Review by antomor — CHANGES_REQUESTED** [review]
Just a test

File: core/bin/zksync_api/src/bin/dev_ticker_server.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/bin/dev_ticker_server.rs:37]: could we update this, so it can configurable?

File: docker/dev-ticker/Dockerfile
Comment (franciscotobar) [inline docker/dev-ticker/Dockerfile:19]: Do we want to copy this, or create a volume? Or both?

File: core/bin/zksync_api/src/bin/dev_ticker_server.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/bin/dev_ticker_server.rs:34]: Nice approach :)

File: core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs:240]: Where are we getting this address?

File: core/bin/zksync_api/src/fee_ticker/validator/types.rs
Comment (franciscotobar) [inline core/bin/zksync_api/src/fee_ticker/validator/types.rs:4]: Im not sure about this, but we may discuss it or I may be confused, but do we need the Clone, derive?

File: docker-compose.yml
Comment (franciscotobar) [inline docker-compose.yml:30]: Im not sure if we need to build the image each time.

Comment (antomor) [reply docker-compose.yml:30]: @franciscotobar this will be changed in your PR, once we merge it.

File: docker/dev-ticker/Dockerfile
Comment (jurajpiar) [reply docker/dev-ticker/Dockerfile:19]: I'm not sure what you mean. here we just copy it from the builder.

File: core/bin/zksync_api/src/bin/dev_ticker_server.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/bin/dev_ticker_server.rs:37]: well that would certainly make sense, however:
1. this is unchanged from the orig impl
2. we might not care what port it is on internally (in docker), instead we probably just want to be able to configure the port binding upon container creation
3. we'd have to add the config to
  1. dev_ticker config
  2. fee_ticker config (coingecko_url)
  4. Dockerfile

File: core/bin/zksync_api/src/fee_ticker/validator/types.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/fee_ticker/validator/types.rs:4]: 1. this is copied from CoinGecko rust sdk (I know that doesn't mean we should keep it)
2. It would be safer, but also more annoying to work with. 

Say we want to pass it to two functions. Ideally, w'd just be passing the reference to it. But if we have to pass it to a function that takes it by value, then we have to clone, or it will be moved to that function. If it isn't returned by this function, then it would be lost when the function ends.

**Review by antomor — DISMISSED** [review]

File: docker/dev-ticker/Dockerfile
Comment (franciscotobar) [reply docker/dev-ticker/Dockerfile:19]: maybe we can do it in a different task, improving the Dockerfile. This copy the tokens file but its better to have a volume with it.

File: core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs
Comment (franciscotobar) [reply core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs:240]: yeah but could we put a comment about it, I think that this is from mainnet inside the `/etc/tokens/`. @antomor this is the same from the other unit tests.

File: core/bin/zksync_api/src/fee_ticker/validator/types.rs
Comment (franciscotobar) [reply core/bin/zksync_api/src/fee_ticker/validator/types.rs]: Thanks for the explanation :)

File: core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs:240]: I don't get the need. It's not like the tests will be ever ran against mainnet

File: docker/dev-ticker/Dockerfile
Comment (antomor) [reply docker/dev-ticker/Dockerfile:19]: In this specific case, I don't see the benefit of using a volume.

File: core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs
Comment (jurajpiar) [reply core/bin/zksync_api/src/bin/providers/dev_liquidity_provider.rs]: I gave it a random address and a comment

File: docker/dev-ticker/Dockerfile
Comment (franciscotobar) [reply docker/dev-ticker/Dockerfile]: By not using a volume, when I run the container without the volume, its going to use the localhost, if it exists, when the image was created. This container its just for local deployment, isn't it?

Comment (jurajpiar) [reply docker/dev-ticker/Dockerfile]: so I removed the copy

Comment (franciscotobar) [reply docker/dev-ticker/Dockerfile]: we can discuss it, after the daily, I will leave it there.

**Review by franciscotobar — APPROVED** [review]
LGTM

**Pull request merged**
